### PR TITLE
[Inventory] Create and review purchase requests + normalize inventoryData

### PR DESCRIPTION
**Pull request by wangannie**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

# What's new in this PR
This PR allows users to:
- Submit purchase requests for specific inventory items
- Approve or deny purchase requests (currently any user can do this, not limited to admins)
Note that this only covers bare functionality. I'll continue to update the front-end as the designs evolve.

**Additional restructuring & cleanup**
Made the following updates inspired by Redux docs ([normalizing state shape](https://redux.js.org/recipes/structuring-reducers/normalizing-state-shape), [createEntityAdapter](https://redux.js.org/tutorials/fundamentals/part-8-modern-redux#using-createentityadapter)):
- Switched from using `mapStateToProps` to  `useSelector`
- Refactored `inventoryData` to normalize using `createEntityAdapter`for inventory updates, products, site inventory, and purchase requests.

**TODO before merging:**

- [x] Update patch for `requests.js`

## How to review
On the backend, run the latest version of `main`.

**Create a purchase request**
1. In the Inventory tab, select an item and click **Purchase**
2. Fill out the form (note: you can't add a receipt yet) and hit **Confirm**
3. This should take you back to the inventory profile. Verify that:
    - The inventory quantity is updated (old qty + purchase request qty)
    - The request was correctly created in Airtable
    - The request was created in Redux with a valid record id if online, or a randomly generated integer if offline. Note: if you're getting an integer id when online, check the console for an error message.
4. Navigate back to the main Inventory screen and select **Purchase Requests** at the top of the list (this is a placeholder for navigation, pending designs). You should see the purchase request you just created at the top of the **Pending Review** section.

**Review the purchase request**
1. In the **Purchase Requests** screen, select a purchase request under **Pending Review** and click either **Approve** or **Deny**.
2. This should take you back to the list of purchase requests where the request you just reviewed will move from the **Pending** section to either the **Approved** or **Denied** section (scroll down).
3. Verify that the update is reflected in Airtable and Redux (`status`, `reviewerId`, and `approvedAt` should all change).

Test this with both Approve and Deny. If you approve the request, it should update the inventory item's current quantity as well (check Airtable and Redux).

**Test redirects**
1. Reset the Redux store
2. Go directly to http://localhost:3000/inventory/purchase-requests/purchase-request by pasting it into the URL bar
3. Since there is no `currentInventoryId` set in state, it should redirect you to the main inventory screen.

**Switching between sites**
1. In the East Asian Library site, navigate to Inventory and select a product (e.g. Diesel) to view the profile page. In the Redux inspector, `currentInventoryId` under `inventoryData` should match the record id of the selected item.
2. Go back to the main page `/home` and switch to a different site. In the Redux inspector, `currentInventoryId` should be reset to an empty string.

**Edge cases**
- After going through this workflow for a site, try changing the site and making sure it still works.

### Online sources
https://react-redux.js.org/using-react-redux/connect-mapstate
https://momentjs.com/docs/#/utilities/
https://www.typescriptlang.org/docs/handbook/enums.html
https://inspiredwebdev.com/find-replace-in-array/
https://redux.js.org/tutorials/essentials/part-6-performance-normalization

### Related PRs

[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

## Next steps
- Admin permissions: limiting the ability to review requests to admins and giving the ability to lookup other user info by id
- Front-end updates to reflect designs
- Combine purchase requests and inventory updates into an inventory history to show on the inventory page
- Add functionality to upload and display receipt images
- Error messaging/input validation

### Screenshots

https://www.loom.com/share/85caa7e71615475c8591e004c2de7674

CC: @julianrkung

[//]: # 'This tags Julian as a default. Feel free to change, or add on anyone who you should be in on the conversation.'

### CODE REVIEW

File: src/lib/redux/inventoryDataSlice.ts
Comment (wangannie) [inline src/lib/redux/inventoryDataSlice.ts]: Not really sure how to handle this -- should I rename to avoid confusion?

File: src/lib/airtable/request.js
Comment (wangannie) [inline src/lib/airtable/request.js]: I originally had this as a custom endpoint (see backend PR) but switched it to this -- do you think this is fine for this use case?

File: src/lib/airtable/request.js
Comment (wangannie) [inline src/lib/airtable/request.js]: I'm not really confident this warrants a custom endpoint. Thoughts?

File: src/lib/redux/inventoryData.ts
Comment (julianrkung) [inline src/lib/redux/inventoryData.ts]: maybe `updateInventoryQuantityInRedux`? Keeps it consistent with `updatePurchaseRequestInRedux` and the function's name now shows what it's updating

File: src/lib/redux/inventoryDataSlice.ts
Comment (julianrkung) [inline src/lib/redux/inventoryDataSlice.ts]: NIT: use siteId instead of `action.payload.siteId` since you already have a var for it

File: src/screens/Inventory/CreatePurchaseRequest.tsx
Comment (julianrkung) [inline src/screens/Inventory/CreatePurchaseRequest.tsx]: Side note (probably not for this PR): We should do a refactor soon to wipe away `any` usage for location. It's preventing us from typing out what's getting passed as props via react-router and therefore makes the code significantly harder to read and maintain IMO.

File: src/screens/Inventory/InventoryProfile.tsx
Comment (julianrkung) [inline src/screens/Inventory/InventoryProfile.tsx]: use the `getProductByInventoryId` you made here

File: src/lib/airtable/request.js
Comment (julianrkung) [reply src/lib/airtable/request.js]: We might also want to set the precedent that "Purchase requests must have been persisted to Airtable before they can be approved" (The only workflow that this would cut off is an admin submitted a purchase request and self-approving it).

We might want to do this because the workflow might get too complicated for the workflow of:
1. Offline inventory created
2. While still offline, a purchase request is created
3. While still offline, that same purchase request (which doesn't have a valid id) gets reviewed

File: src/lib/airtable/request.js
Comment (julianrkung) [reply src/lib/airtable/request.js]: I don't think this warrants a custom endpoint. In the event that we want to do admin authentication (checking the `reviewerId` is an admin, which we should do) on the backend, We can write it in the `write` access resolver for purchase requests.

Details on how to do that with airlock / access resolvers is on: [this](https://share.descript.com/view/mu2V9m33Qa4) demo at around 8:46. If you wanna skip ahead via the transcript, you can CMD+F "So if your rules differ depending on whether or not your users are reading or writing to a base"

File: src/screens/Inventory/PurchaseRequest.tsx
Comment (julianrkung) [inline src/screens/Inventory/PurchaseRequest.tsx]: Jsut so I know: Is the lookup to check if they're admins?

File: src/screens/Inventory/PurchaseRequests.tsx
Comment (julianrkung) [inline src/screens/Inventory/PurchaseRequests.tsx]: abstracting this out into a helper function inside of this component would be good (since you're doing the same thing 3 times, 1 for each pending ,approved, denied).

File: src/screens/Inventory/components/InventoryCard.tsx
Comment (julianrkung) [inline src/screens/Inventory/components/InventoryCard.tsx]: use the `getProductByInventoryId` you made here

File: src/lib/airtable/request.js
Comment (julianrkung) [inline src/lib/airtable/request.js]: TBH don't know if i really like the name `reviewPurchaseRequest`. Maybe `updatePurchaseRequestAndInventory`? I think my main concern is just that "review" doesn't seem like an action that should persist something.

File: src/lib/utils/inventoryUtils.ts
Comment (julianrkung) [inline src/lib/utils/inventoryUtils.ts]: NIT: generally we want verb names. `getLastUpdated` or `calculateLastUpdated` or something along those lines?

**Review by julianrkung — COMMENTED** [review]
Left some comments. I haven't tried out workflow yet. Will do that on next iteration!

File: src/lib/redux/inventoryDataSlice.ts
Comment (julianrkung) [inline src/lib/redux/inventoryDataSlice.ts]: Lets follow the all-caps enum convention we have in codebase already https://github.com/calblueprint/meepanyar/blob/main/src/lib/i18next/translator.ts#L8

File: src/lib/redux/inventoryDataSlice.ts
Comment (julianrkung) [reply src/lib/redux/inventoryDataSlice.ts]: We're probably fine here especially if it's behind a `setter` function. If anything, we could probably use namespaces to deal with it?

File: src/screens/Inventory/PurchaseRequest.tsx
Comment (wangannie) [reply src/screens/Inventory/PurchaseRequest.tsx]: Oh oops lookup here is referring to eventually looking up the user's username, since currently I just have it displaying their ID. I guess yeah this would depend on admin access to be able to see a full list of users?

File: src/lib/redux/inventoryData.ts
Comment (julianrkung) [inline src/lib/redux/inventoryData.ts]: Put the `currentSiteSelector` in the `siteData.ts`

While at it, could you put a TODO to remove the `getCurrentSite` getter (we can move all of `siteData` to selectors / entities in a separate PR)?

File: src/lib/redux/inventoryData.ts
Comment (julianrkung) [inline src/lib/redux/inventoryData.ts]: Triple equal instead of double equal for `inv.id === currentInventoryId`

File: src/lib/redux/inventoryDataSlice.ts
Comment (julianrkung) [inline src/lib/redux/inventoryDataSlice.ts]: maybe `currentInventoryItemId`? Not sure if that's better, but it did take me a little to make sure this pointed to an inventory item. Drawback to this is that the Airtable table is literally called `Inventory`, so i'm fine leaving the way it is. Your call on this one

Current naming convention makes `currentInventorySelector` ambiguous (unclear by name if it should return the entire inventory or the current inventory item).

File: src/lib/redux/inventoryDataSlice.ts
Comment (julianrkung) [inline src/lib/redux/inventoryDataSlice.ts]: Is there a reason that only `inventoryUpdates` is an `EntityAdapter`? Will there be a refactor in the future to make the rest `EntityState`s?

(I'm good with just this structure as long as there'll be a follow-up PR to make the rest `EntityAdapter`s

File: src/lib/redux/inventoryDataSlice.ts
Comment (julianrkung) [inline src/lib/redux/inventoryDataSlice.ts]: NIT: just for defensive code, I'd like to check the index here is not -1 (I'm not sure if that's possible. Ex: Offline requests persists right before this method is called, so `action.payload.id` is an offlineId but `pr.id` has Airtable IDs).

File: src/lib/redux/inventoryDataSlice.ts
Comment (julianrkung) [inline src/lib/redux/inventoryDataSlice.ts]: Same comment as ^. Defensive code, check the index as a safety measure

File: src/screens/Inventory/AddInventory.tsx
Comment (julianrkung) [inline src/screens/Inventory/AddInventory.tsx:96]: I see we're removing `mapStateToProps` in favor of `useSelector`. I'm fine with this, but is there a general rule where we'd want to use `mapStateToProps` over `useSelector`? Or do we want to just migrate to `useSelector` longterm (assuming we're using functional components)

File: src/screens/Inventory/CreatePurchaseRequest.tsx
Comment (julianrkung) [inline src/screens/Inventory/CreatePurchaseRequest.tsx]: We should make sure to restrict these to numbers. I put in letters and it let me submit and the amonut became a `NaN`

image

File: src/lib/airtable/request.js
Comment (julianrkung) [inline src/lib/airtable/request.js]: This is now just `updatePurchaseRequest` right? Since inventory update is in `createPurchaseRequest`

File: src/lib/airtable/request.js
Comment (julianrkung) [inline src/lib/airtable/request.js]: now should be called `createPurchaseRequestAndUpdateInventory`?

File: src/screens/Inventory/PurchaseRequests.tsx
Comment (julianrkung) [inline src/screens/Inventory/PurchaseRequests.tsx:30]: Are we fine passing purchaseRequest here via react-router? I'm good with it, just want to make sure it's intentional. A `currentPurchaseRequestId` might seem annoying to deal with

**Review by julianrkung — COMMENTED** [review]
Left some comments! Code looks a lot cleaner with the selectors, good job :)

File: src/screens/Inventory/CreatePurchaseRequest.tsx
Comment (wangannie) [reply src/screens/Inventory/CreatePurchaseRequest.tsx]: I made it default to 0 if NaN temporarily, but tabling this for a later PR to address app-wide error states and input validation.

File: src/screens/Inventory/PurchaseRequests.tsx
Comment (wangannie) [reply src/screens/Inventory/PurchaseRequests.tsx:30]: Yeah I figured this flow would be fine (at least with the current use cases) since we don't really deal with editing a purchase request (besides reviewing, in which case the reviewer is sent back to the main list anyways). I'll probably re-evaluate once we have more finalized designs and user flows. If you think it's likely we're going to standardize  _everything_ to keeping an currentXXXId in state, I'm happy to refactor this too.

File: src/lib/redux/inventoryDataSlice.ts
Comment (wangannie) [reply src/lib/redux/inventoryDataSlice.ts]: Yeah I definitely understand the confusion here, but I think I'm trying to stick to this standard:
- `siteInventory`: list of inventory items that a site has
- `inventory`: a single inventory 'item', aligning with the Airtable table

So this way, for example, `selectCurrentInventory` represents the current inventory 'item', and `selectCurrentSiteIventory` = list of inventory items at a site.

Let me know what you think. I'm not super attached to my naming scheme so if you think it's better to shift to 'item' I'd be open to considering it.

File: src/screens/Inventory/AddInventory.tsx
Comment (wangannie) [reply src/screens/Inventory/AddInventory.tsx:96]: Yeah I think in general I ended up moving all of them over to use selectors because I was always grabbing the same kinds of info, which made it a lot cleaner and more readable to switch to selectors. Also, it got pretty nasty with all the nested params I needed for typescript to agree with `mapStateToProps` (eg `const mapStateToProps = (state: RootState, ownProps: { location: { state: { purchaseRequest: { inventoryId: string; }; }; }; }) => ({` from [here](https://github.com/calblueprint/meepanyar/pull/60/files/afe73e075a92a0ef25939bb4b0eb687b55748af2#diff-bd8ec877cd98f9458d9ae9715e5c346b6a69661fc3cd5671b522b40f1f4ad27fR61-R63).

In general, I think the Redux docs do a better job explaining this, but they did mention

> The existing connect API still works and will continue to be supported, but the hooks API is simpler and works better with TypeScript. ([Source](https://react-redux.js.org/api/hooks#hooks))

File: src/lib/redux/inventoryDataSlice.ts
Comment (wangannie) [reply src/lib/redux/inventoryDataSlice.ts]: wOoooO this was fixed using `createEntityAdapter`'s built-in magic ✨

File: src/lib/airtable/request.js
Comment (julianrkung) [inline src/lib/airtable/request.js:246]: Can you make a note that this doesn't handle offline yet? In particular I don't believe it handles the flow:
1. An inventory item was created while offline
2. While still offline, a purchase request was created for that inventory item (in this case the inventory item doesn't have an ID)

Let me know if i'm off the mark on that one. We can do that in a future PR

File: src/lib/redux/inventoryDataSlice.ts
Comment (julianrkung) [inline src/lib/redux/inventoryDataSlice.ts]: This entire setup and file looks really nice, really good work :)

I'll work on refactoring `customerData` and `siteData` to look similar to this

File: src/screens/Inventory/PurchaseRequest.tsx
Comment (julianrkung) [inline src/screens/Inventory/PurchaseRequest.tsx]: rm commented line if not needed

File: src/screens/Inventory/PurchaseRequest.tsx
Comment (julianrkung) [inline src/screens/Inventory/PurchaseRequest.tsx]: Is there any way this can be split into 2 lines? The line is 180 characters and is kinda hard to read

Comment (julianrkung) [inline src/screens/Inventory/PurchaseRequest.tsx]: Would you mind doing some redirecting if someone navigates to this URL by typing it in (maybe just redirecting them to inventoryMain)? I'll be doing the equivalent handling in the customer-related pages.

File: src/screens/Inventory/InventoryProfile.tsx
Comment (julianrkung) [inline src/screens/Inventory/InventoryProfile.tsx]: Would you mind doing some redirecting if someone navigates to this URL by typing it in (maybe just redirecting them to inventoryMain) when these aren't set yet? I'll be doing the equivalent handling in the customer-related pages.

File: src/screens/Inventory/CreatePurchaseRequest.tsx
Comment (julianrkung) [inline src/screens/Inventory/CreatePurchaseRequest.tsx]: Would you mind doing some redirecting if someone navigates to this URL by typing it in (maybe just redirecting them to inventoryMain) when these aren't set yet? I'll be doing the equivalent handling in the customer-related pages.

File: src/screens/Inventory/components/PurchaseRequestCard.tsx
Comment (julianrkung) [inline src/screens/Inventory/components/PurchaseRequestCard.tsx]: Can we separate this out to 2 lines? This line is 165 characters. Maybe have getting the currentSiteInventoryId the line before?

**Review by julianrkung — COMMENTED** [review]
Really small NITs at this point.

Couldn't check out the actual flow (only looked at the code) because when I tried to log in using backend `main` I got the following error:
image

**Review by julianrkung — COMMENTED** [review]
The error resolved itself after I cleared site data (devtools --> application --> storage --> clear site data). I checked the workflow and everything looked good on that end. I'll approve after the changes from previous PR are done

File: src/screens/Inventory/CreatePurchaseRequest.tsx
Comment (wangannie) [reply src/screens/Inventory/CreatePurchaseRequest.tsx]: `location` is conveniently phased out with selectors :~)

File: src/screens/Inventory/PurchaseRequest.tsx
Comment (wangannie) [reply src/screens/Inventory/PurchaseRequest.tsx]: Added redirects and updated the PR description with a testable workflow!

File: src/lib/airtable/patches/patch-schema.patch
Comment (wangannie) [inline src/lib/airtable/patches/patch-schema.patch:2]: Not sure if this is necessary/what it's actually doing, but this was generated when I ran the git diff.

**Review by julianrkung — APPROVED** [review]
Followed the review instructions and everything looks good. I also switched sites and went to `/inventory/purchase-requests/purchase-request` to make sure that the `currentInventoryId` was reset and I was redirected to inventory main. Everything looks good :). You're the best! This was built out so quickly and with such high quality I'm excited for future PRs HAHA

We can do this in a future PR, but this Purchase Request column should probably just be called "Reviewed At" since it gets populated even when I deny
image

**Pull request merged**
